packet_forwarder/src/lora_pkt_fwd.c

```diff
@@ -3047,23 +3073,202 @@ static void gps_process_sync(void) {
-    int    i = lgw_gps_get(NULL, NULL, &coord, &gpserr);
+    int    i = lgw_gps_get(NULL, NULL, NULL, &coord);
 
     /* update gateway coordinates */
     pthread_mutex_lock(&mx_meas_gps);
     if (i == LGW_GPS_SUCCESS) {
         gps_coord_valid = true;
         meas_gps_coord = coord;
-        meas_gps_err = gpserr;
         // TODO: report other GPS statistics (typ. signal quality & integrity)
     } else {
         gps_coord_valid = false;
     }
     pthread_mutex_unlock(&mx_meas_gps);
 }
 
-void thread_gps(void) {
+
+/* Get the number of bytes waiting in the GPS's buffer. */
+static int i2c_gps_available(size_t * avail) {
+    uint8_t UBX_BYTES_AVAIL_REG = 0xFD;
+    uint8_t read_buf[2] = {0, 0};
+    struct i2c_msg rdwr_msgs[] = {
+        {
+            .addr = 0x42,
+            .len = 1,
+            .buf = &UBX_BYTES_AVAIL_REG,
+        },
+        {
+            .addr = 0x42,
+            .flags = I2C_M_RD,
+            .len = 2,
+            .buf = read_buf,
+        },
+    };
+    struct i2c_rdwr_ioctl_data rdwr_data = {
+        .msgs = rdwr_msgs,
+        .nmsgs = 2,
+    };
+    if (ioctl(gps_dev_fd, I2C_RDWR, &rdwr_data) < 0) {
+        *avail = 0;
+        perror("WARNING: failed to read GPS available data count");
+        return -1;
+    }
+    uint16_t avail_ = 0;
+    unpack_le_u16(&avail_, read_buf);
+    *avail = (size_t)avail_;
+    return 0;
+}
+
+static int i2c_gps_read(size_t n, uint8_t * dst) {
+    uint8_t UBX_DATA_REG = 0xFF;
+    struct i2c_msg rdwr_msgs[] = {
+        {
+            .addr = 0x42,
+            .len = 1,
+            .buf = &UBX_DATA_REG,
+        },
+        {
+            .addr = 0x42,
+            .flags = I2C_M_RD,
+            .len = n,
+            .buf = dst,
+        },
+    };
+    struct i2c_rdwr_ioctl_data rdwr_data = {
+        .msgs = rdwr_msgs,
+        .nmsgs = 2,
+    };
+    if (ioctl(gps_dev_fd, I2C_RDWR, &rdwr_data) < 0) {
+        perror("WARNING: failed to read from GPS buffer data");
+        return -1;
+    }
+    return 0;
+}
+
+#define DELAY_ON_I2C_GLITCH 50
+
+void thread_gps_i2c(void) {
+    size_t const volatile canary_h = 0xdeadfeed;
+    char read_buf[128]; /* buffer to receive GPS data */
+    size_t const volatile canary_l = 0xdeadfeed;
+    size_t wr_idx = 0;     /* pointer to end of chars in buffer */
+
+    /* variables for PPM pulse GPS synchronization */
+    enum gps_msg latest_msg; /* keep track of latest NMEA message parsed */
+
+    /* initialize some variables before loop */
+    memset(read_buf, 0, sizeof read_buf);
+
+    while (!exit_sig && !quit_sig) {
+        assert(canary_h == 0xdeadfeed);
+        assert(canary_l == 0xdeadfeed);
+
+        size_t rd_idx        = 0;
+        size_t frame_end_idx = 0;
+        size_t n_queued      = 0;
+
+        /* poll gps for count of buffer bytes */
+        while (i2c_gps_available(&n_queued)) {
+            MSG("WARNING: n_avail failed\n");
+            wait_ms(DELAY_ON_I2C_GLITCH);
+        }
```

Comment:
How glitchy is this polling routine percentage-wise? Does it fail 50% of the time? 80% of the time?

Reply:
The glitchiness fades in and out over time. It's hard to say. If I had to guess I'd say it fails about 30% of the time

Reply:
While a failure rate of 30% isn't great I think it's good enough for now as long as the frequency of UBX_NAV_TIMEGPS and UBX_NAV_PVT messages is high enough. We can enable GPS interrupts on hotspot v1.1's IO expander later on if that helps.

Do we have that GPS interrupt GPIO wired on hotspot v1.0? If we don't we'll have to keep polling I2C or revert back to serial for that SKU.

Reply:
> Do we have that GPS interrupt pin wired on hotspot v1.0? 
Nope.

> If we don't we'll have to keep polling I2C or revert back to serial for that SKU.
This doesn't affect V1.0 hotspots since the have SX1301s and this codebase is SX1302 only. Additionally, the GPIO pin could be an optional configuration option. That said, using an interrupt doesn't really help much, if at all.

Reply:
Understood. Polling I2C for GPS works for me.